neo/rawio/intanrawio.py

```diff
@@ -282,66 +284,94 @@ def _get_analogsignal_chunk(self, block_index, seg_index, i_start, i_stop, strea
-        if len(shape) == 2:
-            # this is the general case with 2D (ie in the header-attached format)
+        # This is False for Temperature and timestamps
+        mulitple_samples_per_block = len(shape) == 2
+
+        # In the header attached case the data for each channel comes interleaved in blocks
+        # To avoid unecessary memory access we can calculate the blocks we need to access beforehand:    
+        if mulitple_samples_per_block:
             block_size = shape[1]
             block_start = i_start // block_size
             block_stop = i_stop // block_size + 1
-
             sl0 = i_start % block_size
             sl1 = sl0 + (i_stop - i_start)
-
-        if self.file_format != 'header-attached' and stream_id == 'RHD2000 amplifier channel':
-            sigs_chunk = np.zeros((i_stop - i_start, len(channel_ids)), dtype="int16")
-        else:
-            sigs_chunk = np.zeros((i_stop - i_start, len(channel_ids)), dtype="uint16")
-        for channel_index, channel_id in enumerate(channel_ids):
-            if self.file_format == 'header-attached':
-            # Memmap fields are the channel_ids for unique channels
-                data_chan = self._raw_data[channel_id]
-            elif self.file_format == 'one-file-per-signal':
-                if channel_indexes_are_none:
-                    data_chan = self._raw_data[stream_index][:, channel_index]
-                else:
-                    data_chan = self._raw_data[stream_index][:, channel_indexes[channel_index]]
+        
+        # raw_data is a structured memmap with a field for each channel_id
+        for chunk_index, channel_id in enumerate(channel_ids):
+            data_chan = self._raw_data[channel_id]
+            if mulitple_samples_per_block:
+                sigs_chunk[:, chunk_index] = data_chan[block_start:block_stop].flatten()[sl0:sl1]
             else:
-                if channel_indexes_are_none:
-                    data_chan = self._raw_data[stream_index][channel_index]
-                else:
-                    data_chan = self._raw_data[stream_index][channel_indexes[channel_index]]
+                sigs_chunk[:, chunk_index] = data_chan[i_start:i_stop]
 
-            # if shape is 1 it is either temp a 1d or a 1d array from one-file-per-stream
-            # or one-file-per-channel. In which case we do not want to do the block math
-            if len(shape) == 1 or self.file_format != 'header-attached':
-                sigs_chunk[:, channel_index] = data_chan[i_start:i_stop]
-            else:
-                sigs_chunk[:, channel_index] = data_chan[block_start:block_stop].flatten()[sl0:sl1]
+        return sigs_chunk
+
+    def _get_analogsignal_chunk_one_file_per_channel(self, i_start, i_stop, stream_index, channel_indexes):
+
+        signal_data_memmap_list = self._raw_data[stream_index]
+        channel_indexes_are_slice = isinstance(channel_indexes, slice)
+        if channel_indexes_are_slice:
```

Comment:
What about this:

```python
channel_indexes = slice(1,10)
```

It's not how we currently have the code structured since we can just do a slice without making the indexes into a slice, but in this case this would still trigger grabbing all channels rather than channels 1-10. This is a low probability event, but isn't the easiest to debug, no?

Reply:
I get you now. That makes sense. Let me correct it to take into account the structure of the slice. Very good point.

Reply:
I added code to handle this case.

Reply:
cool. looks good.